tests/integration/test_dirbrowser.py

```diff
@@ -0,0 +1,184 @@
+#
+# qutebrowser is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# qutebrowser is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with qutebrowser.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Test the built-in directory browser."""
+
+import os
+import bs4
+import collections
+
+import pytest
+
+from PyQt5.QtCore import QUrl
+from qutebrowser.utils import jinja
+
+
+class DirLayout:
+
+    """Provide a fake directory layout to test dirbrowser."""
+
+    Parsed = collections.namedtuple('Parsed', 'path, parent, folders, files')
+    Item = collections.namedtuple('Item', 'path, link, text')
+
+    LAYOUT = [
+        'folder0/file00',
+        'folder0/file01',
+        'folder1/folder10/file100',
+        'folder1/file10',
+        'folder1/file11',
+        'file0',
+        'file1',
+    ]
+
+    @classmethod
+    def layout_folders(cls):
+        """Return all folders in the root directory of the layout."""
+        folders = set()
+        for path in cls.LAYOUT:
+            parts = path.split('/')
+            if len(parts) > 1:
+                folders.add(parts[0])
+        folders = list(folders)
+        folders.sort()
+        return folders
+
+    @classmethod
+    def get_folder_content(cls, name):
+        """Return (folders, files) for the given folder in the root dir."""
+        folders = set()
+        files = set()
+        for path in cls.LAYOUT:
+            if not path.startswith(name + '/'):
+                continue
+            parts = path.split('/')
+            if len(parts) == 2:
+                files.add(parts[1])
+            else:
+                folders.add(parts[1])
+        folders = list(folders)
+        folders.sort()
+        files = list(files)
+        files.sort()
+        return (folders, files)
+
+    def __init__(self, factory):
+        self._factory = factory
+        self.base = factory.getbasetemp()
+        self.layout = factory.mktemp('layout')
+        self._mklayout()
+
+    def _mklayout(self):
+        for filename in self.LAYOUT:
+            self.layout.ensure(filename)
+
+    def file_url(self):
+        """Return a file:// link to the directory."""
+        return jinja.file_url(str(self.layout))
+
+    def path(self, *parts):
+        """Return the path to the given file inside the layout folder."""
+        return os.path.normpath(str(self.layout.join(*parts)))
+
+    def base_path(self):
+        """Return the path of the base temporary folder."""
+        return os.path.normpath(str(self.base))
+
+    def parse(self, quteproc):
```

Comment:
Sounds good, it landed there because I thought I'd need to pass the temp dir myself when getting the html content, but as it turns out, quteproc already has such a helper function :wink: